File: src/Game/TiledMap.cpp
#include "TiledMap.hpp"
#include "Utils.hpp"
#include "Core.hpp"

bool TiledMap::load(const std::string &filePath)
{
    for (auto layer : _tileMaps) {
        layer->destroy();
    }
    std::string fileContent = Utils::readFile(filePath);

    if (fileContent.size() <= 0)
        return false;
    _mapContent = Utils::fileContentToJson(fileContent);
    _loadTileSet();
    _loadTileMap();
    return true;
}

void TiledMap::destroy()
{
    std::destroy(_mapContent.begin(), _mapContent.end());
    for (auto tilemap : _tileMaps) {
        tilemap->destroy();
    }
}

void TiledMap::render()
{
    for (auto& map : _tileMaps) {
        map->render();
    }
}

void TiledMap::hide()
{
    for (auto& map : _tileMaps) {
        map->hide();
    }
}

bool TiledMap::_loadTileMap()
{
    auto layers = _mapContent["layers"];

    for (auto layer : layers) {
        int height = layer["height"];
        int width = layer["width"];
        int x = layer["x"];
        int y = layer["y"];
        auto cells = layer["data"];

        TileMap *tilemap = new TileMap(x, y, width *
        _tileWidth, height * _tileHeight, _tileWidth, _tileHeight);
        int index = 0;
        float posX = 0;
        float posY = 0;

        for (auto cell : cells) {
            int cellId = cell;
            if (index >= height) {
                index = 0;
                posY = 0;
                posX += _tileHeight;
            }
            if (cellId > 0) {
                sf::Texture *texture = R_GET_RESSOURCE(sf::Texture, std::to_string(cellId));
                Entity *e = new Entity();

                e->getComponent<Layer>()->set(0);
                e->addComponent<Transform2D>()->position = sf::Vector2f(posY, posX);
                e->addComponent<Sprite>(texture);
                e->addComponent<BoxCollider>(e->getComponent<Transform2D>()->position,
                                             sf::Vector2f(16, 16))->setType(BoxCollider::DYNAMIC);
                tilemap->emplaceEntity(e);
            }
            posY += _tileWidth;
            index++;
        }
        //tilemap->render();
        _tileMaps.push_back(tilemap);
    }
    return true;
}

void TiledMap::_loadTileSet()
{
    auto tilesets = _mapContent["tilesets"];
    for (auto& tileset : tilesets) {
        std::string source = tileset["source"];

        // Handle tileset parsing
        std::string tilesetInfoStr = Utils::readFile("../assets/" + source);
        nlohmann::json tilesetInfo = Utils::fileContentToJson(tilesetInfoStr);

        int imgHeight = tilesetInfo["imageheight"];
        int imgWidth = tilesetInfo["imagewidth"];
        std::string img = tilesetInfo["image"];
        int tileWidth = tilesetInfo["tilewidth"];
        int tileHeight = tilesetInfo["tileheight"];
        int id = 1;

        _tileWidth = tileWidth;
        _tileHeight = tileHeight;
        for (int y = 0; y < imgHeight; y += tileHeight) {
            for (int x = 0; x < imgWidth; x += tileWidth) {
                R_ADD_TILE(std::to_string(id), img, x, y, tileWidth, tileHeight);
                id++;
            }
        }
    }
}

File: assets/Scripting/TiledMap.lua
local json = require "assets.Scripting.json"

local TiledMap = {}
TiledMap.__index = TiledMap

local tileMap = {}
local mapJson = nil

local saveEntity = {}
local saveMap = {}

local tileSize = {
    x=0,
    y=0
}

function displayTable(tbl)
    for key, value in pairs(tbl) do
        print(key, value)
    end
end

function TiledMap.new()
    local self = setmetatable({}, TiledMap)
    return self
end

function TiledMap.load(mapPath)
    for k,v in pairs(tileMap) do
        v:destroy()
    end

    mapContent = Utils.readFile(mapPath, true)
    if mapContent == nil or mapContent:len() <= 0 then
        print("Error: Map not found")
        return false
    end
    mapJson = json.decode(mapContent)
    TiledMap.loadTileSet()
    TiledMap.loadTileMap()
    return true
end

function TiledMap.loadTileSet()
    local tilesets = mapJson["tilesets"]

    for k, tileset in pairs(tilesets) do
        local source = tileset["source"]

        tilesetStr = Utils.readFile("./assets/" .. source)
        tilesetJson = json.decode(tilesetStr)

        imgHeight = tilesetJson["imageheight"]
        imgWidth = tilesetJson["imagewidth"]
        imagePath = tilesetJson["image"]
        tileSize.x = tilesetJson["tilewidth"]
        tileSize.y = tilesetJson["tileheight"]

        local id = 1
        for y=0, imgHeight - 1, tileSize.y do
            for x=0, imgWidth - 1, tileSize.x do
                ResourceManager.addTile(tostring(id), imagePath, x, y, tileSize.x, tileSize.y)
                id = id + 1
            end
        end
    end
    return true
end

function TiledMap.loadTileMap()
    local layers = mapJson["layers"]

    layerIndex = 0
    for x, layer in pairs(layers) do
        height = layer["height"]
        width = layer["width"]
        x = layer["x"]
        y = layer["y"]
        cells = layer["data"]

        tilemap = TileMap.new(x, y, width * tileSize.x,
                height * tileSize.y, tileSize.x, tileSize.y)
        index = 0
        posX = 0.0
        posY = 0.0

        for i, cell in pairs(cells) do
            cellId = cell

            if index >= height then
                index = 0
                posY = 0
                posX = posX + tileSize.y
            end
            if cellId > 0 then
                texture = ResourceManager.getTexture(tostring(cellId))
                e = Entity.new()

                e:getLayer():set(layerIndex)
                transform = e:addTransform2D(posY, posX)
                e:addSprite(texture)
                if layerIndex >= 1 then
                    e:addBoxCollider(transform.position,
                            Vector2f.new(tileSize.x, tileSize.y)):setType(BoxColliderType.DYNAMIC)
                end
                tilemap:emplaceEntity(e)
                table.insert(saveEntity, e)
            end
            posY = posY + tileSize.x
            index = index + 1
        end
        table.insert(saveMap, tilemap)
        tilemap:render()
        layerIndex = layerIndex + 3
    end
end

function TiledMap.hide()
    for k, v in pairs(tileMap) do
        v:hide()
    end
end

function TiledMap.render()
    for k, v in pairs(tileMap) do
        v:render()
    end
end

function TiledMap.drawBox()
    for k, e in pairs(saveEntity) do
        collider = e:getBoxCollider()
        if collider ~= nil then
            Core:draw(collider)
        end
    end
end

function TiledMap.destroy()
    for k,v in pairs(saveMap) do
        v:destroy()
    end

    for k,v in pairs(saveEntity) do
        v:destroy()
    end
end

return TiledMap